src/org/sosy_lab/common/collect/SkipList.java

```diff
@@ -0,0 +1,756 @@
+      boolean existed = subList.add(currNode.getValue());
+      assert !existed;
+      currNode = currNode.getNext(LEVEL_ONE);
+    }
+
+    return subList;
+  }
+
+  @Override
+  public SortedSet<T> headSet(T pToElement) {
+    Preconditions.checkNotNull(pToElement);
+    return subSet(first(), pToElement);
+  }
+
+  @Override
+  public SortedSet<T> tailSet(T pFromElement) {
+    Preconditions.checkNotNull(pFromElement);
+    return subSet(pFromElement, last());
+  }
+
+  @Override
+  public Comparator<? super T> comparator() {
+    return comparator;
+  }
+
+  @Override
+  public Iterator<T> iterator() {
+    return new ListIterator<T>() {
+
+      private Node<T> currentNode = head;
+      private int idx = 0;
+      boolean removed = false;
+
+      @Override
+      public boolean hasNext() {
+        return currentNode.getNext(LEVEL_ONE) != null;
+      }
+
+      @Override
+      public @Nullable T next() {
+        if (!hasNext()) {
+          throw new NoSuchElementException();
+        }
+        // noinspection ConstantConditions
+        currentNode = currentNode.getNext(LEVEL_ONE);
+        assert currentNode != null;
+        idx++;
+        removed = false;
+        return currentNode.getValue();
+      }
+
+      @Override
+      public boolean hasPrevious() {
+        return currentNode.getPrevious(LEVEL_ONE) != null;
+      }
+
+      @Override
+      public @Nullable T previous() {
+        if (!hasPrevious()) {
+          throw new NoSuchElementException();
+        }
+        // noinspection ConstantConditions
+        currentNode = currentNode.getPrevious(LEVEL_ONE);
+        assert currentNode != null;
+        idx--;
+        return currentNode.getValue();
+      }
+
+      @Override
+      public int nextIndex() {
+        if (!hasNext()) {
+          throw new NoSuchElementException();
+        }
+        return idx + 1;
+      }
+
+      @Override
+      public int previousIndex() {
+        if (!hasPrevious()) {
+          throw new NoSuchElementException();
+        }
+        return idx - 1;
+      }
+
+      @Override
+      public void remove() {
+        if (currentNode == head || removed) {
+          throw new IllegalStateException();
+        } else {
+          removeNode(currentNode);
+          // noinspection ConstantConditions
+          currentNode = currentNode.getPrevious(LEVEL_ONE);
+          removed = true;
+        }
+      }
+
+      @Override
+      public void set(T pT) {
+        throw new UnsupportedOperationException(
+            "Setting the value at a specific position " + "doesn't make sense in a sorted list");
+      }
+
+      @Override
+      public void add(T pT) {
+        throw new UnsupportedOperationException(
+            "Adding a node at a specific position doesn't " + "make sense in a sorted list");
+      }
+    };
+  }
+
+  @Override
+  public String toString() {
+    return "SkipList: [" + "size: " + size + " ]";
```

Comment:
The `toString` of collections is pretty much standardized, and should be the same here as well.